### ISSUE TITLE
The minor release of 16.4 causes BREAKING changes in getDerivedStateFromProps

### ISSUE THREAD

**Issue by goyney** [facebook/react#12898]
According to semver, only non-breaking changes are supposed to go into minor and patch version releases. With the release of 16.4, you have made breaking changes to `getDerivedStateFromProps`. Our entire codebase, running perfectly on 16.3.2 is a dumpster fire as soon as we raise that dependency to 16.4.

The only thing in your CHANGELOG about this *breaking* change is:

> Properly call getDerivedStateFromProps() regardless of the reason for re-rendering. (@acdlite in #12600 and #12802)

Please revert this change and save it for 17.0.0, or provide proper documentation to what this change actually entails so that its use can be adjusted by those who have already implemented `getDerivedStateFromProps`.

**Comment by gaearon** [comment]
I’m sorry this broke things for you. Can you show the relevant code please?

We’re aware this change can cause issues in rare cases, as [documented in the blog post](https://reactjs.org/blog/2018/05/23/react-v-16-4.html#bugfix-for-getderivedstatefromprops). Unfortunately *not* making this change also leads to buggy behavior, although less obvious and deterministic so maybe you didn’t bump into it yet. I can see your point that fixing it is “breaking” in your case, but I hope you can also see that fixing *any* bug can be considered breaking for the code that relied on it.

If your code relied on the old behavior but doesn’t work with the new one, it worked by accident. It’s our fault for overlooking this in the 16.3 release but it took about a month for us to discover this problem from bug reports at Facebook. The function takes two arguments: props and state. But it only runs when one of them updates. This didn’t really make sense and was an oversight in the initial implementation.

When we fixed that issue at Facebook (which, as we learned, was the reason behind numerous bugs), one component that relied on the buggy behavior broke (just like in your case). This was just one component out of a thousand or so. We decided that it’s better in the long term to make the fix than to allow more components to be written that rely on the bug.

Again, I’m sorry this caused issues for you. Seeing your code would be helpful.

**Comment by gaearon** [comment]
@chase Can you be more specific as to where you see the contradiction? It fires more often but the examples in the post still work the same way.

**Comment by aweary** [comment]
@gaearon [this section](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#new-lifecycle-getderivedstatefromprops) states:

> The new static getDerivedStateFromProps lifecycle is invoked after a component is instantiated as well as when it receives new props.

**Comment by aweary** [comment]
> Note that React may call this method even if the props have not changed.

I guess I assumed that meant that it may get called if the parent re-renders, but the props passed to the child didn't change. There is some significant ambiguity there though. FWIW I think over-specifying in an RFC is better than under-specifying.

**Comment by goyney** [comment]
I'm not sure I understand the benefit or reasoning to have this method called on every single change. 

We had to go back and write much more logic to handle the internals of all the `getDerivedStateFromProps` method calls after this change that feels more like boilerplate code the framework should be (and was before 16.4) handling. It just add more noise and bloat to our components.

The 16.3.2 version acted more in lines with the former `shouldComponentUpdate`, which this method replaced. In the pre-16.4 behavior, it was much easier to migrate from the deprecated `shouldComponentUpdate` to the new `getDerivedStateFromProps`.

The blog post provided by @chase states:

> The new static getDerivedStateFromProps lifecycle is invoked after a component is instantiated as well as when it receives new props.

This is exactly what we expected and how it operated in 16.3. That is much cleaner in my opinion.

As a side note, and on a personal level, I never liked the deprecation of `shouldComponentUpdate` and have complained about it internally to many of my coworkers. However, I understand the side-effects that are trying to be edged out by changing methodologies on how the lifecycle works. That being said, if the React group feels that the new behavior introduced in 16.4 is how they wish it to work, I would like to suggest another lifecycle method is introduced more in line with how `shouldComponentUpdate` worked, or undeprecating that method and putting more protections around it to try and edge the issues you intended to prevent. Of course, I am not an architect, nor an employee of Facebook, or a member of the React development team, but the changes around the lifecycle in this particular area are extremely frustrating as a developer.

**Comment by gaearon** [comment]
If we want to continue this discussion let's start talking about specific code examples.
If you have code that got broken by this, please show it and let’s talk about it.

**Comment by gaearon** [comment]
>The 16.3.2 version acted more in lines with the former shouldComponentUpdate, which this method replaced. In the pre-16.4 behavior, it was much easier to migrate from the deprecated shouldComponentUpdate to the new getDerivedStateFromProps.

Sorry, but this seems like a significant confusion on your side. **`shouldComponentUpdate` was never deprecated.** It also has nothing to do with what `getDerivedStateFromProps` is for.

**Comment by gaearon** [comment]
Also I'd like to ask everyone to keep in mind that we're not doing changes just to piss people off. We want what is best for apps in longer term, and we care about not adding extra boilerplate as deeply as you do. We'll make another blog post about this next week. It seems like more people use `getDerivedStateFromProps` than we expected, and that is likely a misunderstanding about its purpose. We'll show more appropriate strategies that don't introduce the kind of boilerplate you're concerned with.

But to continue this discussion **we need code examples**. You can see otherwise it immediately turns abstract and vague.

**Comment by goyney** [comment]
@gaearon No, no confusion, just poor copy/pasting. I immediately edited it. Thanks.

**Comment by gaearon** [comment]
Could you post an example of the code that got broken by the new behavior?

**Comment by goyney** [comment]
Here is the _simplest_ example I can provide from our morning of refactoring to support the update.

**Before:** (only called on prop changes)
```
static getDerivedStateFromProps(nextProps) {
    analyticsPageview(nextProps.documentId);

    return {
      source: nextProps.source,
      selection: {
        startBlockIndex: 0,
        startOffset: 0,
        endBlockIndex: 0,
        endOffset: 0
      }
    };
  }
```

**After**: (refactored because it is called every change)
```
static getDerivedStateFromProps(nextProps, prevState) {
    if (prevState.documentId !== nextProps.documentId) {
      analyticsPageview(documentId);

      return {
        documentId: nextProps.documentId,
        selection: {
          startBlockIndex: 0,
          startOffset: 0,
          endBlockIndex: 0,
          endOffset: 0
        },
        source: nextProps.source
      };
    }

    return prevState;
  };
```

At this point, the `documentId` never will change. Previously, there was no reason to include it in the state. Additionally, we needed to wrap an additional logic statement around the entire thing. This specific example is small and I can see the "no big deal" attitude that some people may give based on it but again, we have much larger refactors that this change caused.

**Comment by gaearon** [comment]
I believe the “before” code already had a bug (so 16.4 just surfaced it).

>At this point, the documentId never will change.

Right, but the parent component could re-render (e.g. if it had `setState` call itself, or if one if its parents did). In that case `getDerivedStateFromProps` would also get called twice—which you consider a bug because it triggers analytics twice.

That's exactly the kind of broken assumption that led to bugs at Facebook, and led us to do this change.

Note how in the 16.3 blog post [we explicitly demonstrates you need to keep previous values in the state](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#updating-state-based-on-props), just like you ended up doing. Then you wouldn't have this problem now.

Additionally, `getDerivedStateFromProps` is intended to be a pure method. It’s **not** an appropriate place for side effects (use `componentDidUpdate` for them instead). I’m sorry if this wasn’t clear from the docs—we’ll update them if it wasn’t.

**Comment by chase** [comment]
For what it's worth, none of my code broke after the update. It was specifically based off of [this example on the blog](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#fetching-external-data-when-props-change).

Based on the changes mentioned in the newest post, I had assumed that this would no longer behave as expected:
``` javascript
  static getDerivedStateFromProps(nextProps, prevState) {
    // Store prevId in state so we can compare when props change.
    // Clear out previously-loaded data (so we don't render stale stuff).
    if (nextProps.id !== prevState.prevId) {
      return {
        externalData: null,
        prevId: nextProps.id,
      };
    }

    // No state update necessary
    return null;
  }
```

**Comment by gaearon** [comment]
I don’t see why that example would break with the update. Am I missing something? As far as I can tell the conditional check would be false on state-only updates, and it would return null.

**Comment by RoyalHunt** [comment]
Why I still have an error: http://i.imgur.com/fjFSDgq.png , when I do not have any `componentWillReceiveProps` in my project: http://i.imgur.com/cXl7Kad.png ?

**Comment by chase** [comment]
I suppose the fact it looked so similar to what the blog post states breaks is what threw me off:
``` javascript
static getDerivedStateFromProps(props, state) {
  if (props.value !== state.controlledValue) {
    return {
      // Since this method fires on both props and state changes, local updates
      // to the controlled value will be ignored, because the props version
      // always overrides it. Oops!
      controlledValue: props.value,
    };
  }
  return null;
}
```

The similarity is superficial:
``` javascript
static getDerivedStateFromProps(nextProps, prevState) {
  if (nextProps.id !== prevState.prevId) {
    return {
      // Since this method fires on both props and state changes, local updates
      // to the controlled value will be ignored, because the props version
      // always overrides it. Oops!
      prevId: nextProps.id,
    };
  }
  return null;
}
```

**Comment by benwiley4000** [comment]
The given explanation for why the change was made is reasonable, it seems, but that's a separate issue from the fact this change was released without a major semver bump. 16.4 breaks what was generally understood to be the correct behavior in 16.3.

I'm glad none of my own code broke, but for a project as widely adopted as React it seems reckless to ship a breaking change under the guise of a "bugfix." I understand it may be inconvenient and out-of-step and bad marketing to release a new major version all of a sudden, but on the flipside it makes React's versioning less reliable and it makes me a bit more hesitant about upgrading to new React versions.

**Comment by RoyalHunt** [comment]
Actually I have some breaking problems as well with using apollo 1. But I solved this with the next code:
`static getDerivedStateFromProps(nextProps, state) {
    const savedArticles = nextProps.data.articles
    if (savedArticles && !state.firstRender && savedArticles !== state.savedArticles) {
      return {
        savedArticles,
        firstRender: true
      }
    }
    return null
  }`

**Comment by codeaid** [comment]
@gaearon I don't have any code examples (and I've not even installed 16.4 yet) but I just wanted to comment on what you said earlier...

> It seems like more people use `getDerivedStateFromProps` than we expected, and that is likely due to a misunderstanding about its purpose. For example, it might be that the code that relied on `componentWillReceiveProps` is always getting updated to `getDerivedStateFromProps` whereas memoization or lifting state up would have been more appropriate.

I'm surprised you're surprised about this. I fully expected people to use this method simply because core methods on which, I'm sure, many relied upon previously are being removed. Namely those that allow us to tell when props change externally. I, personally, got the impression that this is almost the new `componentWillReceiveProps` from reading the docs.

As for me, my use case for `getDerivedStateFromProps` are components that receive objects to be edited from external sources (usually from Redux via `mapStateToProps`) and store that data in the local state (which has to be done as you can't edit property values directly). They also usually some `onSave(newData)` type callbacks (mapped via `mapDispatchToProps`) that are triggered when the updated data needs to be persisted. Those callbacks result in new data being pushed into the current component's props and local state having to be updated.

Before `getDerivedStateFromProps` existed I had to check for presence of and changes to `this.props.something` in both - `constructor` and `componentWillReceiveProps` - because data could be available during the construction, or alternatively it might've been necessary to load it first, meaning it would become available through one of the eventual `componentWillReceiveProps` calls.

`getDerivedStateFromProps` is obviously a godsend in that regard and makes the code so much cleaner and neater and I personally can't see any other way to implement behaviour similar to what I described above. It would definitely be interesting to see how _memoization or lifting state up would have been more appropriate_ as you said. I'm pretty sure I looked into those options but couldn't find a solution that would work for these use cases. `componentDidUpdate` can be used in some occasions but sometimes it's too late and things need to be checked before something gets changed and not after.

**Comment by bvaughn** [comment]
> I hear that there was some intentional ambiguity in the spec, but that doesn't change how folks understood gDSFP to work.

Our original implementation was flawed. For what it's worth, the [RFC spec for `getDerivedStateFromProps`](https://github.com/reactjs/rfcs/blob/master/text/0006-static-lifecycle-methods.md#static-getderivedstatefrompropsnextprops-props-prevstate-state-shapestate--null) did make mention of this:
> Note that React may call this method even if the props have not changed. If calculating derived data is expensive, compare next and previous props to conditionally handle changes.

But we failed to community the warning clearly on the API reference docs, and I apologize for that.

---

> I, personally, got the impression that this is almost the new `componentWillReceiveProps` from reading the docs.

We need to do a better job of communicating when `getDerivedStateFromProps` is appropriate and when other techniques would be better. (See [this tweet for an example](https://twitter.com/brian_d_vaughn/status/987443099537309696).) We plan to publish another blog update in the next week or two that covers this in more detail.

> It would definitely be interesting to see how _memoization or lifting state up would have been more appropriate as you said_.

We will be sure to include several examples in the upcoming update about this.

**Comment by bvaughn** [comment]
That's a good point, @codeaid. It's can be difficult to strike the right balance between too much and too little detail, but I think our current reference section definitely needs some work.

We hope to eventually create a "recipes" section on the site that shows common tasks and our suggested patterns. Unfortunately there are only a couple of us though, and this sort of thing is pretty time consuming.

But I appreciate the feedback. We'll try to keep it in mind as we make small edits over the next couple of weeks.

**Comment by codeaid** [comment]
@bvaughn Ask Mark to give you more people! It's ridiculous that a library this popular globally is getting so few resources :)

**Comment by benwiley4000** [comment]
Assign Mark a few code reviews at least!

**Comment by dubbha** [comment]
Am I the only one who now just can't grasp the naming and the param names of this method?
It's called `getDerivedStateFromProps` and the params are `(nextProps, prevState)`, and it totally made sense in 16.3 - we were literally getting derived state from props, on the new props, thus having nextProps and prevState as params. Yeah, it was kind of a new componentWillReceiveProps, but at least the naming made sense, particularly when I was going through this [diagram](https://twitter.com/dan_abramov/status/981712092611989509) two days ago.
Now when the same method will fire on setState() and forceUpdate(), I am totally confused. Why do I need it on setState()? How do I get the newState in this case? How is a prevState useful in this case? What am I deriving from what? And why do I need it on forceUpdate()?

**Comment by gaearon** [comment]
Would it help to say that arguments are just called `props` and `state` and represent current props and state? 

I thought prefixes would be helpful but I see now they’re more confusing.

**Comment by bvaughn** [comment]
No, you're not the only one, @dubbha!

Andrew and I were talking about this earlier this week and decided to update the docs to just be `props` and `state` like in the [most recent blog example](https://reactjs.org/blog/2018/05/23/react-v-16-4.html#2-compare-incoming-props-to-previous-props-when-computing-controlled-values) but I guess neither of us have made this change yet. I will do it now!

**Comment by gaearon** [comment]
@RoyalHunt

>Actually I have some breaking problems as well with using apollo 1.

Looking at your “before” code in https://github.com/facebook/react/issues/12898#issuecomment-391847365, it has the exact same bug as I described in https://github.com/facebook/react/issues/12898#issuecomment-391782913. 

So here, too, React 16.4 didn’t break your code. It helped you find the bug that already existed (previously unnoticed) in your code and manifested itself less deterministically. 


Sounds like a good thing to me.

**Comment by Tolgeros** [comment]
Am I correct in understanding that an updated version of [this diagram](https://twitter.com/dan_abramov/status/981712092611989509) would have gDSFP stretching to the right, such that both setState and forceUpdate will now also trigger it?

**Comment by dubbha** [comment]
@gaearon, @bvaughn, first of all thank you for your answers.

> Would it help to say that arguments are just called props and state and represent current props and state?

Yes, it helps, at least it is more obvious that what I am getting as the second argument in case of `setState()` is the current, newState, not the prevState instead.

> The use cases haven’t changed. If you needed this method in 16.3 and your code doesn’t have bugs (such as the one described above), it should also work in 16.4. So if you “need it” it’s for the same reason you “needed it” in 16.3.
> If you don’t need it that’s cool too—it’s not intended to be commonly used. As I already noted a few times in this thread it’s hard to say whether you need it if you don’t show a particular snippet of code.

When asking the "why do I need it" question I was mostly trying to think of the new use cases for the method. It covers the `new props` use cases perfectly. I was thinking of a use case for the `setState()` and `forceUpdate()`. There were no such use case before of course, so there's no code to show and nothing is broken. I am just not getting how this method could be useful in the latter two cases. Just can't think of the use cases. And what I would be deriving from what.

@Tolgeros 
>Am I correct in understanding that an updated version of this diagram would have gDSFP stretching to the right, such that setState will now also trigger it (but not forceUpdate)?

Nope, it also fires on forceUpdate() (but skips shouldComponentUpdate):
https://codesandbox.io/s/wk7qxnvk38

**Comment by andrewBalekha** [comment]
For example, in this case, `getDerivedStateFromProps` triggers in each `setState`, thus I have always default value, which is prop value (15).

```javascript
class RangeInput extends React.Component {
  state = {
    value: this.props.value,
  }

  static propTypes = {
    value: PropTypes.number.isRequired,
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.value !== prevState.value) {
      return {
        value: nextProps.value,
      }
    }

    return null
  }

  onChangeRange = e => {
    this.setState({ value: parseFloat(e.currentTarget.value) })
  }

  render() {
    const { value } = this.state

    return (
      <div>
          <input
            type="range"
            min="0"
            max="100"
            step="0.5"
            value={value}
            onChange={this.onChangeRange}
          />
        </div>
    )
  }
}

ReactDOM.render(<RangeInput value={15} />, document.querySelector("#app"))
```

is here another workaround?

**Comment by gaearon** [comment]
@Tolgeros in https://github.com/facebook/react/issues/12898#issuecomment-391946524

>Also, would this be a sensible way, in general, to mimic the 16.3 behavior in 16.4?

Yes, I think this looks equivalent to me. I don’t recommend writing code like this though. It is fragile. Let me expand on that.

>It needs a copy of 'value' kept in its local state, for user experience reasons, but if changes come from above it needs to be able to respond to it.

Just to clarify we’re on the same page: do you realize that just re-rendering parent for any reason whatsoever counts as “change from above”?

For example, if you wrap your component in some popup container that passes `closePopup` as a prop to its child, and that function is different on every render (so equality checks in Redux containers won’t stop it), and the popup calls `setState`, your component will re-render and blow away the user input.

I think any code that tries to distinguish “change” from above by comparing whole props objects instead of individual props (or, in the past, just relying on when the method gets called) will lead to bugs like this.

To implement something like you described, it would make sense to me if you at least determined the change by comparing some explicit marker coming from Redux store (e.g. an ID or version field) that explicitly gets incremented when you *intend* to blow away the local state.

Do you see what I mean?

@dubbha
            
>Yes, it helps, at least it is more obvious that what I am getting as the second argument in case of setState() is the current, newState, not the prevState instead.

Well yes, the only reason we called it `prevState` in the docs is because you are *returning some state already*. So the input state is definitely “older” than what you are creating right now. I was worried people would attempt to mutate the state object if it was just called `state`.

>When asking the "why do I need it" question I was mostly trying to think of the new use cases for the method.

Maybe I’m wrong but I don’t think the change is intended to cover *new* use cases. It is intended to uncover dormant bugs in *existing* use cases, as we’ve seen already three times in this thread:

* https://github.com/facebook/react/issues/12898#issuecomment-391847365
* https://github.com/facebook/react/issues/12898#issuecomment-391782913
* And the issue described in this comment

So the use cases are the same, and if you use the method in supported ways, you don’t need to make adjustments. If your code has bugs, they became more prominent, and you need to fix the code to follow the supported patterns (like in the documentation and blog posts).

@andrewBalekha
            
>For example, in this case, getDerivedStateFromProps triggers in each setState, thus I have always default value, which is prop value (15).

Your code also has the same problem as in three cases above that I described, even in React 16.3.

If the user inputs something but then parent component re-renders for a different reason, you will blow away user input (because it won’t match the state).

For every component you need to decide if some value is **controlled** or **uncontrolled**. This is part of your component’s API design.

If it is **controlled** then you should use the value from props, and not attempt to mirror it in state. When you want to change something, you need to *ask the parent* to do so. This is called [“lifting state up” and is described in detail in documentation](https://reactjs.org/docs/lifting-state-up.html). In this scenario you **don’t need `getDerivedStateFromProps`**. 

If it is **uncontrolled** then you can take a prop like `defaultValue` for the first render. But from that point you should use local state to keep the value, and *re-rendering from above shouldn’t reset it*. In this scenario you **don’t need `getDerivedStateFromProps either`**. 

Combining controlled and uncontrolled behavior (using `getDerivedStateFromProps` or other means) is possible but extremely error-prone. I strongly suggest against it. You can see from my three previous bug descriptions in this thread that this leads to the same mistake: a completely unrelated parent re-render will blow away your state.

So what do you do instead if you *want* to blow away the state when something changes in the parent? **You need to figure out what exactly should serve as a trigger for blowing away the local state.** For example maybe in some parent there is a state like `currentItemID` that changes. In that case, you can either:

* Pass *that state* down to the child so it can use it in `getDerivedStateFromProps` and decide to blow away the existing local state.
* Or you can give `key={this.state.currentItemID}` to the child so that when the ID changes, it is forced to be unmounted and then mounted again with new props. This will ensure the “clean slate” without `getDerivedStateFromProps` but will blow away existing DOM. And in fact that might be exactly what you need if conceptually the “current item” has changed and you don’t want to preserve things like focus, uncontrolled input state, or scroll positions.

Does this explain the problem a little bit? I’m not sure if my explanations are clear enough because I keep describing the same problem in the application code but every new example in this thread demonstrates it again as if it was desired behavior.

**Comment by andrewBalekha** [comment]
@gaearon, thanks I've got my problem now.

Here is an example where my code is failing in 16.3 as well https://codesandbox.io/s/j809y1llv

**Comment by karolk** [comment]
Simpler example of the issue: https://jsfiddle.net/6fme1y69/2/
It seems like coming from componentWillReceiveProps in 16.2 to getDerivedStateFromProps in 16.3 it was very easy to refactor the code to something that stopped working properly in 16.4.

**Comment by gaearon** [comment]
@karok 

This already didn’t “work properly” in React 16.3. I believe I explained in detail why this code is buggy in several comments above. Did you get a chance to read them yet?

I’m currently away from keyboard but I can create a demo fiddle if my comments are unconvincing.

**Comment by bvaughn** [comment]
@karolk Dan already mentioned why this wouldn't work in a few comments above. Essentially it boils down to:

> if the user inputs something but then parent component re-renders for a different reason, you will blow away user input (because it won’t match the state).

For your example, even in React 16.3, if the component re-rendered because another prop changed (e.g. an `onChange` type callback), then your `state` would be overridden. This mixing of `props` and `state` has awkward/broken edge cases, even before `getDerivedStateFromProps`.

I'm going to quote from [Dan's comment right above](https://github.com/facebook/react/issues/12898#issuecomment-392035163) because it's a great summary:

---

> For every component you need to decide if some value is **controlled** or **uncontrolled**. This is part of your component’s API design.

> If it is **controlled** then you should use the value from props, and not attempt to mirror it in state. When you want to change something, you need to *ask the parent* to do so. This is called [“lifting state up” and is described in detail in documentation](https://reactjs.org/docs/lifting-state-up.html). In this scenario you **don’t need `getDerivedStateFromProps`**. 

> If it is **uncontrolled** then you can take a prop like `defaultValue` for the first render. But from that point you should use local state to keep the value, and *re-rendering from above shouldn’t reset it*. In this scenario you **don’t need `getDerivedStateFromProps either`**. 

> Combining controlled and uncontrolled behavior (using `getDerivedStateFromProps` or other means) is possible but extremely error-prone. I strongly suggest against it. You can see from my three previous bug descriptions in this thread that this leads to the same mistake: a completely unrelated parent re-render will blow away your state.

**Comment by karolk** [comment]
@gaearon I think my example is slightly different. I updated the fiddle to demonstrate the behaviour I was going for. I was trying to have a box where you can type or paste anything but only a value meeting certain criteria ends up in a store and comes back as a prop. I am aware that this can be achieved differently and it is maybe using gDSFP in a way that's too similar to cWRP. https://jsfiddle.net/6fme1y69/3/

**Comment by karolk** [comment]
@bvaughn I went for a combination of parent controlled and state but this seems to depend on implementation detail. I don't think there's anything I don't get in @gaearon's explanation. Thank you both for your time, I appreciate quick response here and on Twitter.

**Comment by bvaughn** [comment]
@karolk Did you have a chance to read my reply or the sections of Dan's reply that I highlighted?

It seems like you're aiming for an **uncontrolled** component, in which case we strongly recommend that you don't try to mirror `props` in `state` like your component is doing. Instead, you can initialize `state` with a default value from `props` like so:

```js
class Test extends React.Component {
  constructor(props) {
    super(props);

    // Read default text from incoming props (store),
    // But then don't try to mirror it to state.
    this.state = {
      text: props.defaultText
    };
  }

  changeText(newText) {
    this.setState({ text: newText }, () => {
      // Notify store of text change if some criteria is met.
      if (this.state.text.match(/\d+/g) === null) {
        dispatch(this.state.text);
      }
    });
  }

  render() {
    return (
      <input
        onChange={({ target }) => this.changeText(target.value)}
        value={this.state.text}
      />
    );
  }
}
```

Even for React 16.2 (before there was `getDerivedStateFromProps`) mirror `props` and `state` in the way your fiddle shows was not recommended. (If it would be helpful for me to elaborate on why this is, I could show an example or two.)

**Comment by gaearon** [comment]
I’m going to close this. In conclusion:

* If this change exhibits bugs in your code in React 16.4, **your code was already buggy in React 16.3 in more subtle and less deterministic ways**.
* To support this claim, I refer you to five examples posted in this thread https://github.com/facebook/react/issues/12898#issuecomment-391781205, https://github.com/facebook/react/issues/12898#issuecomment-391847365, https://github.com/facebook/react/issues/12898#issuecomment-391946524, https://github.com/facebook/react/issues/12898#issuecomment-392001463, https://github.com/facebook/react/issues/12898#issuecomment-392048949 **which are all already buggy in React 16.3** as I explained in my analysis in https://github.com/facebook/react/issues/12898#issuecomment-391782913 and https://github.com/facebook/react/issues/12898#issuecomment-392035163.

If you’re not convinced by my explanation, let’s take a demo from https://github.com/facebook/react/issues/12898#issuecomment-392080509.

Indeed, at the first glance seems like [in the version using React 16.4 you can’t edit anything](https://jsfiddle.net/6fme1y69/2/), but [in the version using React 16.3 the code runs fine](https://jsfiddle.net/5wr67hbx/). So is the original code free of bugs?

No, it’s not. Here is [a fiddle using React 16.3 where we re-render the parent component every second](https://jsfiddle.net/0yne6jzp/). You can see that the input resets every second:

![gif of component resetting](https://i.imgur.com/IfaP5sv.gif)

That clearly wasn’t intended. And all examples in this thread that got broken by React 16.4 already have the same exact bug. Whenever a parent re-renders, the state gets unintentionally reset. And it’s pretty hard to debug because components are decoupled, and whether you see this bug or not depends on how “lucky” you are.

React 16.4 helped surface this bug in your code so you can fix it, instead of keeping the app broken in cases that reproduce less reliably.

As we mentioned, we’ll follow up with another post that describes when (not) to use `getDerivedStateFromProps`. I already shared some thoughts on component API design part of this in https://github.com/facebook/react/issues/12898#issuecomment-392035163.

But we maintain that **making an existing bug in your app reproduce more consistently is not a breaking change**, and is a helpful way to help you find and fix that bug. We are sorry for not being clear enough in the documentation and our guidance for using this method, and we’ll amend that by posting more in the coming week.

Finally, I agree this method is kinda hard to wrap your mind around. I don’t think it’s the flaw of this API, it’s just that state-based-on-props was always hard to think about but this complexity was “hidden” due to imperative nature of `componentWillReceiveProps`. So even if bugs like this existed in your apps you might not have been able to reliably reproduce them every time.

The new method makes the complexity of what you’re trying to do explicit. The verbosity helps you realize that perhaps it’s not a good solution for your particular problem, and being more clear whether a value is *controlled* or *uncontrolled*, and lifting state up when necessary, will be better both from readability and correctness perspectives. Again, we’ll follow up with more guidance on that.

Thanks to everyone for the discussion! I hope this was helpful, and if I missed anything please let me know.

**Comment by goyney** [comment]
You broke semver, plan and simple. I understand your explanations and I understand your reasoning and do not disagree with them. But, when I `yarn` and get the latest minor release, my stuff shouldn't stop working because of a dependency change, plan and simple.

While the changes are good intended, that is my major sticking point in all this.

**Comment by sophiebits** [comment]
@goyney Every bug fix has the potential to break your code if you were relying on the buggy behavior. It's true that this bug fix is a little more likely than most to break code, but there isn't a hard line between "bug fix" and "breaking change"; it's much more of a spectrum. When we looked at this we felt it was closer to a bug fix.

**Comment by gaearon** [comment]
Semver is a social contract. I think it’s not as clear cut as it may seem.
Here’s code that may randomly break and then “unbreak” on every single React release:

```js
if (ReactDOM.render.toString().length > 330) {
  throw new Error('oops');
}
```

Does it mean we can never change `render()` implementation because any possible change would break semver? Probably not.

It’s a silly example of course. What about code like this?

```js
<div onPointerCapture={() => { throw new Error('no'); }} />
```

Technically it doesn’t do anything in React 16.3, but throws an error in browsers that support in React 16.4. Does that mean that adding support for pointer events was a breaking change?

This code also looks silly although maybe this version is less far-fetched:

```js
const handlers = {
  onPointerCapture() {
    const x = somethingThatReturnsNull();
    x.foo();
  }
}

// somewhere
<Foo handlers={handlers} />

// ...
//  in Foo:

<div {...this.props} />
```

I think it’s still implausible but we can argue that for any kind of change, there is a chance that there exists some code that happens to rely on the old behavior. If we wanted to respect semver in the strictest (and almost mathematical) sense then we’d have to make every single release major. Which of course defeats the purpose of semver.

So, coming back to my original point, semver is a social contract. We try hard not to break things that weren’t broken. But given the choice between “break already broken code more strictly and ensure more broken code doesn’t get written” and “keep broken code working in some cases, but allow more broken code to be written” it’s less clear cut. In this case, we decided that it’s worse for the ecosystem to endure the pain of fixing it later (and in the meantime live with bugs that are hard to reproduce) than make existing hidden bugs apparent. I can see arguments for doing it the other way but I hope you can see ours too.

**Comment by codeaid** [comment]
I've read most of the comments in this thread but I'm still not quite sure if I'm understanding the planned changes correctly.

From what I understand instead of only being triggered when external properties are changed, `getDerivedStateFromProps` will now get triggered on every external AND internal update. Am I right? if that's the case then what is the alternative for actually only listening for external changes without having to compare props and state all the time?

You're saying if there were bugs in 16.3 they will surface in 16.4, however, I don't see how the use case I described in my [comment](https://github.com/facebook/react/issues/12898#issuecomment-391858156) earlier is covered or even achievable.

Just to recap my problem - there a top level components that "talk" to Redux and get state data pushed into them via props (using `mapStateToProps`) when Redux receives new data. Because these components are pretty much top level components there is no way for me to lift the state meaning I have to accept props AND have them stored in the local state so that I can edit them. Therefore the requirement is to be able to tell when something changed externally so that I can replace my local state. I don't care if I blow user data away because that's exactly the point - to replace local state with fresh data from the API.

I suppose we could say that in a way new props coming in from Redux can be compared to the parent component rerendering - you blow away internal state, however, that is the whole point in these components I have - you edit stuff, trigger an action, make an API call and get new data from Redux that has to replace the current state. Oftentimes `componentDidUpdate` is too late in the lifecycle so that's no good but if `getDerivedStateFromProps` will now get triggered on any change then I'll never be able to tell when something is _about to_ change.

I encountered this problem when updating components from `componentWillReceiveProps` to `getDerivedStateFromProps` but with the way the latter one worked it was a logical change. Now I'm not even sure how to proceed.

**Comment by gaearon** [comment]
>you edit stuff, trigger an action, make an API call and get new data from Redux that has to replace the current state

How do you determine that data is “new”? I don’t understand this part.

Judging by your description, you don’t compare any IDs or anything. So you’re just hoping that your components never get re-rendered unless something specific happens on the Redux side (e.g. a data fetch). This is a very fragile assumption. As far as I can tell adding a single always-changing value to `mapStateToProps` (e.g. `something: {}`) will cause your component to always re-render on *any action* (not just the ones you care about), and always reset the state. Isn’t that bad?

In other words, **you seem to be relying on a performance optimization (React Redux currently bails out of rendering shallowly equal props) for correctness.** This sounds like it will become a bigger problem later on (e.g. if React Redux stops using shallow equality checks everywhere).

It’s hard for me to say more without seeing the code but that’s the basic idea. How do you solve it? I believe I answered in https://github.com/facebook/react/issues/12898#issuecomment-392035163. If you need this kind of behavior, you need some explicit marker that tells you the data is actually “new” (e.g. ID or version field is different) so it needs to be replaced. **Your components should be resilient to being re-rendered.**

**Comment by goyney** [comment]
Semvar is a social contract and it is a contract a company like Facebook should probably be following, especially with such a huge community of developers utilizing a, dare I say, standard in front-end development at this point in time.

You had functionality in a non-alpha/non-beta release out in the wild for two months. You had [cited documentation](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#new-lifecycle-getderivedstatefromprops) stating the lifecycle's behavior. You made claims of `componentWillUpdate` deprecation and suggested developers start moving to the new modal as soon as possible to prepare for the eventual removal of the deprecated code.

With all of this, I would assume this means you had proper unit and contract testing for each piece of the lifecycle, including `getDerivedStateFromProps`. I can almost bet that at some point during 16.4's development tests were added or changed because the behavior changed here. In fact, you can see test file changes in both https://github.com/facebook/react/pull/12802 and https://github.com/facebook/react/pull/12600.

**That should have been a red flag that a breaking change was being made.** A change to the fundamental lifecycle of React is a big deal.

Was it really a bug? Yeah, I think based on the discussions in this thread and your explanations, it should be classified as an unintended behavior. But this wasn't a launch-day slip up. This was _two months_ into the release of 16.3. This is a **breaking change**. It warranted more of a heads-up to the community or it should have been saved for 17.0. This one change wasted the collective community probably several hundred hours of productivity time to figure out why their code just stopped working. I know it took two developers at least two hours to track down why everything just quit working since this specific change omitted no errors.

I also understand that sometimes accidents happen, when a breaking change is pushed unintentionally. But everything you've stated in this thread indicate that this was intentional and you knew what you were doing. To me, that is extremely unprofessional and good way to upset your base of users. Something like this, especially if it happens more than once, can easily push someone to a competing framework.

Listen, I like React. It took a long time for me to ramp up on it and feel really comfortable with it.  But I like it. I'm not going to go bashing it because of stuff like this. It's small, it's petty, and frankly, I have code I need to be writing and arguing with people over the interwebs is not something I like to do. 

What I am asking you is to be more vigilant on how and when you release breaking changes in the future.

**Comment by gaearon** [comment]
Thanks for elaborating! I think we don’t disagree that much. This rings true to me:

>It warranted more of a heads-up to the community

We’ll take this feedback to the heart.

I stand by saying that this is a poorly communicated bugfix for an (also poorly communicated) feature rather than a semver-major change. But we could definitely have explained it better in retrospect, and it’s our fault.

For what it’s worth we *did* mention in the blog post that unsupported patterns that worked in some cases (but not others) by accident would break more consistently.

I see why you’re saying it was annoying to debug, but it’s not like we made this change silently. In fact a large part of the text in 16.4.0 blog post is dedicated to this exact change. If you saw breakage after upgrading, it would make sense to search your code for `getDerivedStateFromProps` (and patterns described in the post) rather than try to find a bug for two hours without that context. I’m not sure whether this is what happened, and I’m definitely not saying it’s your fault, but we did make effort to explain what issues you may run into.

Above all I appreciate everyone’s willingness to engage in this conversation. Again, we’re sorry we let the original version stay out for enough time for you to start depending on the buggy behavior. In fact the biggest reason we scrambled to get 16.4 out sooner was because we realized the mistake a few weeks ago (and we’ve been testing the impact of the fix for these past two weeks, which gave us confidence that the change is scoped enough and the existing bugs it uncovered were important enough to ship the fix now).

**Comment by goyney** [comment]
@gaearon We didn't knowingly upgrade. I wasn't even aware 16.4 had dropped until someone on my team said, "Hey, look at the React version, it changed."

Our company's policy is `^`, so we automatically got the latest version on our next fresh pull of the code, which just happened to be that same night. Under normal circumstances, we may not have discovered the issue for a few days. By that point in time, I'm sure one of us would have saw the release announcement and would have attacked things in a different way.

**Comment by kohlmannj** [comment]
@gaearon Hi! I have a real-world use case (albeit in a nascent library) which breaks in React 16.4 (was on 16.3.2). As I write this I am investigating a refactor of my code; so my goal is simply to provide you with an example of something that broke.

In [kohlmannj/limina](https://github.com/kohlmannj/limina), I've implemented a custom 2D [ScrollView component](https://github.com/kohlmannj/limina/blob/6585d018b3039daab706a1983f8a0357c5e272a1/packages/react-graph-editor/src/components/ScrollView/ScrollView.tsx) (sighhh don't ask). Here's a [demo a working version of the component](https://kohlmannj.github.io/limina/?selectedKind=react-graph-editor%2FScrollView&selectedStory=with%20resizeable%20kitten), which uses React 16.3.2 as of this writing.

Try clicking and dragging the scroll bar caps, [as shown in this tweet](https://twitter.com/jkohlmann/status/993010103820455936). That interaction corresponds to [one](https://github.com/kohlmannj/limina/blob/6585d018b3039daab706a1983f8a0357c5e272a1/packages/react-graph-editor/src/components/ScrollView/ScrollView.tsx#L383) or [another](https://github.com/kohlmannj/limina/blob/6585d018b3039daab706a1983f8a0357c5e272a1/packages/react-graph-editor/src/components/ScrollView/ScrollView.tsx#L385) call to `this.setState()` within `onBeginDrag()`, a higher-order method (sic?) whose returned event handler is attached to [`window.mousemove`](https://github.com/kohlmannj/limina/blob/master/packages/react-graph-editor/src/components/ScrollView/ScrollView.tsx#L299).

I have also implemented [this `getDerivedStateFromProps()` static method](https://github.com/kohlmannj/limina/blob/6585d018b3039daab706a1983f8a0357c5e272a1/packages/react-graph-editor/src/components/ScrollView/ScrollView.tsx#L24), which serves the following purpose:

- The component exposes `this.props.scaleX` and `this.props.scaleY`, which are used to [derive initial `this.state.scaleX` and `this.state.scaleY` values](https://github.com/kohlmannj/limina/blob/6585d018b3039daab706a1983f8a0357c5e272a1/packages/react-graph-editor/src/components/ScrollView/ScrollView.tsx#L38)
- Incoming number-type props changes to `this.props.scaleX` or `this.props.scaleY` should replace the corresponding `this.state` value
- This supports the need for interactively rescaleable scroll view which can also be [re-]set by the parent component

Now, I'm sure you can guess where this is going. In theory this can indeed be fixed by [lifting state up](https://reactjs.org/docs/lifting-state-up.html), which I will certainly consider in my refactor.

—

Here's how I discovered the breakage in my code with React 16.4: I upgraded both `react` and `react-dom`, [re-]started Storybook, navigated back to the "with resizeable kitten" story, and found the component no longer responded to dragging the scroll bar caps. (Here's a reference commit: https://github.com/kohlmannj/limina/commit/c430a909c5a48936aa669cd3b7d16935c474f81d) It seems as though the call to `this.setState()` within the `window.mousemove` event handler now has no effect, since `getDerivedStateFromProps()` is now being called whenever the component updates. Sure enough: when I temporarily commented out `getDerivedStateFromProps()`, I found that the component reacted to dragging the scroll bar caps once again.

—

Anyway, as mentioned, I will look into refactoring my code; this is not a major inconvenience. I hope the use case and explanation thereof helps you and the team better understand how `getDerivedStateFromProps()` might be used. I feel good about the evergreen recommendation to "lift state up", mind you.

**Comment by klimashkin** [comment]
Probably `getDerivedStateToProps` could be renamed to just `getDerivedState` to stop confusing it with `componentWillRecieveProps` and emphasize that it doesn't depends on props change only.

**Comment by kohlmannj** [comment]
@klimashkin Interesting. It's true that I went into my 16.3 updates thinking that `getDerivedStateFromProps()` was more or less a drop-in replacement for `componentWillRecieveProps()`.

Rather than a rename, though, perhaps the React docs could instead emphasize the need to holistically consider a component's state management strategy when transitioning away from `componentWillReceiveProps()`. It might also be good to point out that `getDerivedStateFromProps()`, by nature of its updated behavior in React 16.4, should be used only when the component _exclusively_ derives its state from props, rather than from other interactions, such as event handlers.

**Comment by Dema** [comment]
@kohlmannj For me it is not a bug in react in any case, it's just these two pieces of code are just not equivalent
```javascript
public componentWillReceiveProps(nextProps: IScrollViewProps) {
    if (nextProps.scaleX !== this.props.scaleX && typeof nextProps.scaleX === 'number') {
      this.setState({ scaleX: nextProps.scaleX });
}
```
and 
```javascript
 public static getDerivedStateFromProps(nextProps: IScrollViewProps, prevState: IScrollViewState) {
    if (nextProps.scaleX !== prevState.scaleX && typeof nextProps.scaleX === 'number') {
      return { scaleX: nextProps.scaleX };
}
```
In cWRP you are comparing old prop value with the new one and in gDSFP you are comparing it with derived value, that is going to change inside you component. I think  now with gDSFP you just have to save the old scaleX value in state, too, to be able to compare new prop against it. Am I correct?

```javascript
if(nextProps.scaleX !== prevState.oldScaleX && typeof nextProps.scaleX === 'number') {
    return {oldScaleX : nextProps.scaleX, scaleX : nextProps.scaleX}
}
```

**Comment by kohlmannj** [comment]
@Dema Ah, I see what you mean! Thanks for your analysis. I implemented this change: https://github.com/kohlmannj/limina/commit/7df272930cc2a9bbf227c2fbc94782433ac6f81f

**Comment by catamphetamine** [comment]
React 16.4 broke my code too.
See details: https://github.com/facebook/react/issues/12912

**Comment by gaearon** [comment]
@catamphetamine 

I already asked you to refrain from name-calling about a month ago in another thread. Please stop this.

I am not “blaming” other people for their code being buggy. I’m sorry if I came across this way. I am pointing out that **the bug you’re running into already existed in your code and is not a new one**. I have analyzed it in detail in this thread, including providing a demo.

**This is not a value judgment about you or your code. If we want this discussion to be fruitful, I think it might help to separate the technical details from emotions.**

The same is true for your snippet in https://github.com/facebook/react/issues/12912 with React 16.3. Your phone input would completely reset if `onChange` callback doesn't immediately set the same state, and the parent component re-renders for another reason. If you turn this into a fiddle I’ll be happy to demonstrate that.

Again, I’m sorry that this bug was surfaced by the change (and it frustrated you) but **it already exists in the code you wrote**. I’m not trying to blame you, or make you feel bad. I’m only stating the fact about the code. And, as I mentioned earlier, **we don’t consider making existing bugs in your code occur more often to be a breaking change**.

But we definitely could do better at communicating *how* to use `getDerivedStateFromProps` without introducing bugs, and the blog post we plan to write will hopefully help that.

Peace.

**Comment by catamphetamine** [comment]
@gaearon Ok, thx, I removed my comment. I have been heard which means we're not being opressed here. I had to double-check that.

**Comment by gaearon** [comment]
Thanks. Btw I'd be happy to look at the issue in your input component but no earlier than Tuesday (due to holidays).

**Comment by bvaughn** [comment]
@kohlmannj Regarding [the pattern you described above](https://github.com/facebook/react/issues/12898#issuecomment-392235392), I wanted to reiterate a concern in case it's helpful. (This isn't exactly related to `getDerivedStateFromProps` but more about the overall pattern.)

You described your component as:
> - The component exposes `this.props.scaleX` and `this.props.scaleY`, which are used to [derive initial `this.state.scaleX` and `this.state.scaleY` values](https://github.com/kohlmannj/limina/blob/6585d018b3039daab706a1983f8a0357c5e272a1/packages/react-graph-editor/src/components/ScrollView/ScrollView.tsx#L41)
> - Incoming number-type props changes to `this.props.scaleX` or `this.props.scaleY` should replace the corresponding `this.state` value
> - This supports the need for an interactively rescaleable scroll view which can either be used in "uncontrolled", standalone form; or, in the future*, as controlled by a parent component

My concern is that mixing **controlled** and **uncontrolled** behavior leads to a confusing API and/or confusing runtime behavior. At a high level, it's not clear what the source of truth is when `props` and `state` values disagree. More practically speaking, there are two variants this pattern usually takes- and each has downsides.

### Variant 1: state is always reset to props when the component is re-rendered

The recent `getDerivedStateFromProps` change highlights one downside of this approach, but the problem also existed for version 16.3 as well as older versions built on top of `componentWillReceiveProps`.

The problem is this: **Unexpected re-renders override your component's state unintentionally**. This includes changes to unrelated `props` like callback functions or inline styles. These are often re-created inline during render, and so bypass `PureComponent` or `shouldComponentUpdate` purity checks.

### Variant 2: state is only reset when props value changes

This pattern compares incoming props to previous props and only updates state when the value changes externally. (This is done to avoid the unexpected re-render problem I mentioned above.)

However this also has a downside: **There's no good way for the parent to reset a property to the previous value.**

Here's an example scenario:
1. A `ScrollView` component like yours is used to render a lot of form fields and underneath of them, there's a "submit" button.
1. A user fills out the fields and tries to submit, but there's an error with one field.
1. The application renders `<ScrollView scrollTop={posOfErrorField} {...props} />` to automatically scroll to the field with an error so the user can see it.
1. The user modifies the field, then scrolls down and submits again.
1. There's still an error with this field.
1. The application renders `<ScrollView scrollTop={posOfErrorField} {...props} />` to scroll back to the problematic field so the user knows it still has an error.
1. Nothing happens, because `posOfErrorField` hasn't changed.

This would be confusing for both the _user_ and the _developer_ trying to debug this problem. And the only work arounds are:
* Set a new `key` to recreate the component entirely. (This works fine for a form field example but can be expensive for components that contain a lot of children, like our `ScrollView`.)
* Re-render twice, once with e.g. a null prop and then the second time with the original prop- in order to force the inner component to reset its state.

---

We recommend designing your components to be either fully controlled or fully uncontrolled to avoid the above cases.
* A fully **controlled** component avoids both of the above problems, because there's no state to be reset.
* A fully **uncontrolled** component can be lighter weight to use (since it generally requires fewer callbacks). In this case, props can be used to initialize _default_ state values- but we recommend against using props to _update_ state values for reasons like the ones I've outlined above. In this case, we suggest prop names that are clear like `defaultScrollTop` or `initialScrollTop` etc. If you need the ability to forcefully _reset_ an uncontrolled component, you could add an instance method (to be called via a `ref`).

**Comment by kohlmannj** [comment]
@bvaughn Excellent notes — thank you very much. I was considering a change to fully uncontrolled in https://github.com/kohlmannj/limina/commit/d68e23c7bcb8cbbb872fd4d714848e5826673ed1. Based on this discussion, I will work on entirely separating controlled and uncontrolled behavior as I continue development.

**Comment by gaearon** [comment]
@bvaughn Thanks for writing this. I love how this thread helped crystallize these problems in my head. I’ve been vaguely aware of them but it was hard to articulate what exactly was broken in such APIs. Looking forward to the blog post!

**Comment by catamphetamine** [comment]
@bvaughn I think I have a legitimately valid example of a non-"broken" code that still is neither **controlled** nor **uncontrolled**. And I claim that's it's not "wrong architecture" or anything.

So, I have a phone number input component.
It has a country select with a flag icon and a phone number input field.
The `value` should be able to be set externally at any point in time, and when it does the country select flag must update accordingly.

So, say, one tries to make such a component an **uncontrolled** one.
In this case `value` is taken from `props`.
And the country flag is stored in `state` and must be contstantly re-verified against `props.value`.

```js
static getDerivedStateFromProps(props, state) {
  const { value } = props
  const { country } = state
  const newCountry = getCountryForValue(value)
  if (newCountry && newCountry !== country) {
    return { country: newCountry }
  }
  return null
}
```

So, I'll be computing `getCountryForValue(value)` on each keystroke.
Suppose it's a computationally-heavy function which slows down performance.
Not the best architecture in my opinion.
Scientifically it's all good (purely stateless) but when it comes to practice it's wasting resources (CPU, RAM).

So when you want to make the phone input be performant on mobile devices in order for users to not get "janky" and "buggy" input with poor UX you have to resort to the good-old "stateful" approach.
I haven't fully figured it out yet but I already know that I have to go this way: UX matters, you know, more than having a clean code or whatever. Ultimately that's UX that's generating revenue and making the actual value for humanity.

So, to recap, I'm still experimenting with my code and I will make it work surely, but going fully into a stateless mode may seem like a perfect solution in theory and being bounded just by the React library itself but in the wild there are cases when it just doesn't work.

If you think otherwise you can answer.
But don't write lengthy comments.
I won't be wasting anyone's time on this, I can and will figure it all out myself.

**Comment by catamphetamine** [comment]
Actually, abort.
Don't answer to my comment.
I just had an idea.
I'm already doing `getCountryForValue()` in `<input onChange/>`.
So I might as well just memoize that function with a limit of two arguments.
The first argument will be the "old" value and the second argument will be the "new" value.
When it's inside `<input onChange/>` it will calculate `getCountryForValue()` for the "new" value, so when it steps into `getDerivedStateFromProps()` it will calculate `getCountryForValue()` with the same "new" value (now being taken from `props`) and so it will be a zero-cost operation.
Ok then, I'll be changing my component into a stateless one maybe.
Looks like memoization really saves the "pure-functional" approach performance.

**Comment by bvaughn** [comment]
@catamphetamine The example code you show above looks more like a [**controlled** component](https://reactjs.org/docs/forms.html#controlled-components) to me. (It does not store the current phone number "value" in state; it gets it from props.)

In the example you show above, you don't actually need to use component state at all. The core of what you're doing is deriving something (which flag to show) from props in a way that performs well. All you need to do this sort of thing is a memoization wrapper.

> If you think otherwise you can answer.
But don't write lengthy comments.

👍

**Comment by gaearon** [comment]
I want to add a note to https://github.com/facebook/react/issues/12898#issuecomment-392266811 before I forget about it again.

>### Variant 2: State is only reset when props value changes

>This pattern compares incoming props to previous props and only updates state when the value changes externally. (This is done to avoid the unexpected re-render problem I mentioned above.)

>However this also has a downside: There's no good way for the parent to reset a property to the previous value.

Here is another scenario where this is a bad solution. Often people want to keep edits to a form in local state like `state.value` (and “commit” them to some place above, such as a Redux store, on an event like button click).

Let’s say we have a UI like this:

```
[button: View Previous]
[button: View Next]

Edit value: [input: Value]

[button: Save] 
```

You can navigate items with Previous and Next buttons (and the text input should reset to the current values for them), and you can click Save to save your edits.

To implement this, people sometimes try to use `componentWillReceiveProps` or (now) `getDerivedStateFromProps`. The logic goes like this: we initialize `state.value` to `props.value`. And if `props.value` changes “from above“, we want to throw away local edits (we *assume* the change from above comes due to Previous or Next click). However, this logic is buggy.

Consider the case where we navigate between two items with `Previous` and `Next` buttons. If their values are different, sure, a re-render will reset the field.

But what if two neighbor items have the same value? Then the condition you’re using to determine it (something like `props.value !== state.previousValueFromAbove`) will be false, and **you’ll keep local edits even though the current item has changed, and they’re not valid anymore!**

On the other hand, if your condition looks like `props.value !== state.value` (and compares to current input value), apart from your code breaking in 16.4, in 16.3 you would *also* get buggy behavior I described in earlier comments (every parent re-render would blow away your state).

So how *can* you make this pattern work? Really, the best way is to avoid it, and instead explicitly make the component fully controlled (so that the state can be managed by the component above).

But if you *must* do something like this, you can accept a prop like current item ID, and keep the previous ID in state. Then in `getDerivedStateFromProps()` you know **the current _item_ has changed** and you need to reset the state. Just having the current _value_ changing from above doesn’t really tell you enough to make a decision.

**Comment by sophiebits** [comment]
> So how *can* you make this pattern work? Really, the best way is to avoid it, and instead explicitly make the component fully controlled (so that the state can be managed by the component above).

> But if you *must* do something like this, you can accept a prop like current item ID, and keep the previous ID in state. Then in `getDerivedStateFromProps()` you know **the current _item_ has changed** and you need to reset the state. Just having the current _value_ changing from above doesn’t really tell you enough to make a decision.

Another option is to not implement getDerivedStateFromProps at all – meaning your child component won't reset when you move to the next item – but then specify a `key` on the child that the parent determines using the ID of the item. That will mean that when the key/ID changes, the child is completely remounted. Note that this may be worse for performance in some cases but is quite possibly the semantics you want. (It depends on the use case though.)

Essentially, it boils down to this: When you use getDerivedStateFromProps (or componentWillReceiveProps), your component is *guessing* at what has happened based on the old and new values. Even if you store the old and new values and compare them properly (as we've been suggesting in this thread), you're making a guess at what the parent wanted you to do. In many cases this won't cause problems – for any parent–child pair, it may be that the child will always guess correctly because the parent may only change its state in certain ways. But for a fully reusable child, it's not a good idea to use these methods. Because when you pull them into a new parent that behaves differently, they might do the wrong thing.

**Comment by gaearon** [comment]
Here is a few smaller examples (with a counter) if someone prefers to see the code.

Before (using `componentWillReceiveProps`): https://codepen.io/ismail-codar/pen/gzVZqm?editors=1011

Demonstration of why this code is *already has a bug*: https://codepen.io/anon/pen/jxgXgK?editors=1011. Click “demonstrate bug” and you’ll see that completely unrelated `setState` in parent blows away the child state.

A correct solution that needs neither `componentWillReceiveProps` nor `getDerivedStateFromProps`, and instead lifts state up to the parent so it can decide what to do in either case: https://codepen.io/anon/pen/erqXpz?editors=1111

**Comment by klimashkin** [comment]
I don't think there are some huge problems with controlled/uncontrolled handling in `getDerivedStateFromProps`, it's all solvable.

But on the other side I don't understand purpose of `getDerivedStateFromProps` if you say 'just lift your state up' (I assume it doesn't matter if that upper state holder is another component or redux selector). If you follow 'just lift your state up', you don't need local component state anywhere at all. And you don't need `getDerivedStateFromProps` ever as well. Then react team should make components stateless and thisless (get rid of classes) to force that approach. If last is an option that react team is considering now (and `getDerivedStateFromProps` is an early bird), just let us know so we can change mindset in advance.

Also the quote from docs '`getDerivedStateFromProps` is invoked right before calling the render method' not quite correct, because there is shouldComponentUpdate in between, it will make people think they can move derive logic directly into render method (where they have access to `this` again).

**Comment by gaearon** [comment]
@klimashkin 

>Then react team should make components stateless and thisless (get rid of classes) to force that approach. If last is an option that react team is considering now (and getDerivedStateFromProps is an early bird), just let us know so we can change mindset in advance.

While we’re always considering alternatives to classes, [as we have done for the past four years](https://github.com/reactjs/react-future), **local component state is an essential feature of React** (pretty much *the* most important one), and it’s not going away anywhere. I think you might be taking my comment to an extreme I didn’t intend.

>But on the other side I don't understand purpose of getDerivedStateFromProps if you say 'just lift your state up' (I assume it doesn't matter if that upper state holder is another component or redux selector). If you follow 'just lift your state up', you don't need local component state anywhere at all. 

I didn’t mean you should always lift it up as high as possible. I meant that it’s bad to have duplicate sources of truth that disagree with each other about any particular value. Lifting state up can solve that problem: basically, it forces you to choose one source of truth, and get the data flowing down from there. I didn’t say “lift *all* the way up“, just enough to solve the practical issue.

Even if you lift it up to their shared ancestor, it’s still local state in the component that owns it. But if you have two local states in two components that really are supposed to represent the same conceptual value, and “syncing” them is a pain, this is a sign that lifting state up to their common ancestor (but no higher than necessary) can resolve the ambiguity and the bugs it causes.

Lifting state up isn’t a panacea either. I’m only suggesting it as a solution when you’re literally trying to sync two values. I think that if a prop and state field are named the same way, it’s a good rule of thumb that `getDerivedStateFromProps` might not be an ideal solution because it’s just a symptom of a duplicate source of truth.

But there are legit cases for `getDerivedStateFromProps` that are *not* solved by lifting state up.

Sometimes state really is both *local* and *derivative*. For example, you may want to reset displayed suggestion list when the user changes a field. The suggestions themselves might be an implementation detail of a component, and you might not want to lift them up (thus causing every parent to manage them). I think this would be a legitimate case for using `getDerivedStateFromProps`: you reset some local state in response to a prop change, and the parent doesn’t need to be aware the derived value even exists.

If I were to come up with a checklist for when using `getDerivedStateFromProps` makes sense, it would be something like:

* The derived state is truly local to the component: it doesn’t directly map to any data the parent is already managing.

* The state is only re-derived on specific prop changes. Re-rendering a parent at arbitrary times doesn’t blow away the child state.

* Effectively, this means one of two options. Either this state is accumulated from prop changes (e.g. `scrollDirection` state accumulated from changes to `scrollTop` props). Or it’s accumulated “per prop value” (e.g. autosuggest items displayed for current field, which need to reset whenever the field changes).

**Comment by EyalPerry** [comment]
@gaearon 

> Here is another scenario where this is a bad solution. Often people want to keep edits to a form in local state like state.value (and “commit” them to some place above, such as a Redux store, on an event like button click).

I happen to be working on something very similar, and do have some additional points to list.
I would love your input as I believe there's something you did not address in your example of the single input form in the comments above.

Envision an app, which, upon navigation to some route, loads a json representation of some Form, 
a model, if you will- and then renders this model using React.
Along with the model, data is fetched from the API so that the generated UI is populated with it.
The app enables the user to edit this form and to send it back to the server. Dead Simple.

While the model is fetched only upon navigation, the data can change at any given point due to server side pushes.

Both data and model come in as props from the Redux store,
and since I do not want to dispatch an action to the store on each key stroke,
changes are cached locally in the form component's state, then validated and integrated with the data upon submission.
 
So, with your suggestion- that indicates that child components of said form should be controlled,
The render method of the said form would be invoked whenever a child changes.
This operation, albeit light thanks to react, is quite redundant, since it was not the model that changed.
It was just a value on a single input element. Adding to that, this is no task to repeat unnecessarily on mobile devices, or worse- IoT devices with a UI.

So, in order to try and avoid this, as long as sCU is treated as more than just a hint, the form component only re-renders when either the model or the data have changed- an event which could only originate from a change to store state.

This forces each element to handle it's own state changes, and notify the form of these changes. This enables single elements to re-render, without affecting their siblings.

This, in turn, creates a tiny redundancy since changed element state is duplicated,
but that is something I am willing to live with.

Each element's gDSFP method is implemented such that their initial value is stored in their local state,
So the when the next call to this method comes, the local state is reset only if the value differs from the initially passed value, which only happens if the server decided that it should change.

Is there something I missed, that makes each element susceptible to the bug you described?
Is this implementation considered "problematic" in regard to compatibility with future react versions?
Is there another way to implement said form without causing the form to re-render (albeit virtually) on each element change?

Sincerely
Eyal Perry

**Comment by gaearon** [comment]
>Is there something I missed, that makes each element susceptible to the bug you described?

Yes, any change that would make Form re-render more often (e.g. if it subscribes to some Redux piece of state that updates independently of the input, or has its own local state) will reset the inputs. You may not have this bug right now but it *is* very fragile to rely on this not ever happening.

**Comment by EyalPerry** [comment]
Thanks for taking the time to reply.

I have indeed put sCU on the stateful form elements. guess I was not clear about that.
The cost stems from 'parsing' the same json model into the final result of a createElement call, despite the fact that in most cases, except for the initial render, all but one bail.

About your second remark.. well, what you described is the desired and expected behavior of said Form.
Store wins.  App is designed to overwrite local state due to the user's conscious choice.

Either way, I understand from your words that the React way, as of version 16.4 is that gDSFP should be implemented to be idempotent and that the only risk free practice is to lift state up.

Again, thanks a lot.

**Comment by gaearon** [comment]
>The cost stems from 'parsing' the same json model into the final result returned by the Form.render call, despite the fact that in all cases, except for the initial render, all but one bail.

You can memoize it and then there's no extra cost. For example using https://github.com/alexreardon/memoize-one.

>About your second comment.. well, what you described is the desired and expected behavior of said Form.

I understand your point that form state should win. But imagine you add a *different kind of state* to global store that form cares about (but that's not necessarily this form's state). You wouldn't want that *other* state to reset the form state.

In other words, it's generally expected that *adding* something to `mapStateToProps` result doesn't break the application.

> the only risk free practice is to lift state up

Not necessarily, but it's the one that gives you the most flexibility. We'll publish a blog post soon with an overview of different approaches and their pros and const.

**Comment by EyalPerry** [comment]
Thanks, I will have a look. That blog post will be most welcome.
Best regards, and thanks for taking the time.

**Comment by evoyy** [comment]
@gaearon 

> While we’re always considering alternatives to classes, as we have done for the past four years, local component state is an essential feature of React (pretty much the most important one), and it’s not going away anywhere. 

Thanks for stating this, it reassures me a lot. I'm not a functional purist - I use local state a lot, combined with MobX observables. Just know that there are people like me out there :-)

**Comment by gaearon** [comment]
FWIW we don't think local state is at odds with being functional. But that's a topic for another day.

**Comment by onethreeseven** [comment]
@gaearon 

For context, I am professionally a backend developer but have recently been learning Javascript and React for a personal project.  So if the verdict is that I don't know what I'm talking about, it's probably because I don't know what I'm talking about.  :smile: 

I have a relatively simple case which works in 16.3 and is failing in 16.4, and in addition to hoping you can shed some light on what I'm doing wrong I hope my newbie experience can inform your upcoming post.

The component consists of a radio button and text inputs:
```
() I would like to donate financial aid: $[    ]
() I would like to request financial aid: $[    ]
() I decline to participate in financial aid
```
The state that is useful externally is a single value, which may be positive, negative, zero, or "invalid" (i.e. they selected one of the top two values but didn't input a number) which I represent as `NaN`.  I get the impression that controlled components are preferred, and I do occasionally have to reset the value from above, so I implemented this as a controlled component.

But the state of the component is not fully determined by the controlled value: if the user selects one of the top two radio options, before entering a number the value is `NaN` in either case.  The parent only needs to know that the input is in an invalid state, so to make it cleaner I hid this detail inside the component.

In 16.3 I deal with this using `getDerivedStateFromProps()`, approximately:
```javascript
static getDerivedStateFromProps({value}, {radioState, textState}) {
  if (value !== computeExportedValue(radioState, textState)) {
    return defaultState(value);
  }
}
```
On user input I update the state and call the `onChange` prop if the exported value has changed.  When `getDerivedStateFromProps()` is called it sees that the two match and nothing happens.  (Likewise, upon some unrelated re-render the two continue to match and nothing happens.)

In 16.4 I'm seeing a call to `getDerivedStateFromProps()` with the _old_ value and _new_ state, so it returns the default state for the old value, meaning the component can never change.

If I've accidentally elided anything important, the actual code is [here](https://github.com/onethreeseven/hrsfans-retreat/blob/c200098a7ede560ef8367e43e99de73eafb116ec/static/main.js#L767).

Cheers!

**Comment by bvaughn** [comment]
Hi @onethreeseven 

I've been working on a blog post this week (that I hope to publish today) that covers this topic in more detail, along with examples of what to do and not what to do (and explanations about why). Hopefully you'll read that blog post once it's published. (I'll comment here with a link as well once it's live.)

The code you've shown above would be expected to break in 16.4 because of the changes made to `getDerivedStateFromProps`. Those changes were made to help expose existing (but inconsistently reproducible) bugs.

In your case, you could quick-fix your component by explicitly tracking the input props to see when they change.

```js
static getDerivedStateFromProps(props, state) {
  // Only reset state when the external prop changes.
  // Don't reset it anytime props and state disagree though,
  // Or you will override your own setState() updates.
  if (props.value !== state.prevPropsValue) {
    return {
      prevPropsValue: props.value,
      ... computeExportedValue(state.radioState, state.textState)
    };
  }
  return null;
}
```

**Comment by onethreeseven** [comment]
Okay, thanks!  I look forward to the post.

For posterity, I managed to fix my problem much later last night by changing from (schematically)
```javascript
handleInput(update) {
  setState(update, this.props.onChange(...));
}
```
to
```javascript
handleInput(update) {
  this.props.onChange(...);
  setState(update);
}
```
I had been doing the former because it was slightly more convenient to have the fully updated state on hand when computing the value to pass upwards, but the latter seems to work just as well and only triggers the one expected call to `getDerivedStateFromProps()`.  I hope this turns out to be kosher in the end!

**Comment by bvaughn** [comment]
Here's the blog post!

https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html

**Comment by tankgod** [comment]
![image](https://user-images.githubusercontent.com/11535593/42274450-4f3a8a68-7fbf-11e8-904e-887984bca51a.png)

**Comment by gaearon** [comment]
As explained both in [API reference for `componentDidUpdate`](https://reactjs.org/docs/react-component.html#componentdidupdate) and [blog post](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#side-effects-on-props-change), you should always make a comparison in `componentDidUpdate` to decide whether to do something.

```js
componentDidUpdate(prevProps, prevState) {
  if (prevProps.something !== this.props.something) {
    // do something
  }
  if (prevState.somethingElse !== this.state.somethingElse) {
    // do something else
  }
}
```

**Comment by bvaughn** [comment]
> I tried using ‘getDerivedStateFromProps’ and ‘componentDidUpdate’, but it went into an infinite loop.

Please show what you tried.

**Comment by tankgod** [comment]
I have solved this problem, thank you anyway.

**Comment by lvl99** [comment]
I'm still a bit perplexed on this myself.

While I understand the idea of "controlled" and "uncontrolled", I've got a few places where I want it to be both.

I have two examples (both have the same code, one using React 16.3 and the other React 16.4) with two intentions:
* Control if a button is selected on or off (`ButtonSelectable`)
* Change an object's contents based on form inputs (`EditObject`)

React 16.3: https://codesandbox.io/s/9lopjp809o
React 16.4: https://codesandbox.io/s/rrxjpn7p6m

In both intentions I want to make a copy of the external prop value and save it internally. Equally, I want any external changes to overwrite the internal state. The `ButtonSelectable` is a silly example — in most cases this should always be controlled — however in the `EditObject` one, I want to store a copy of the object in the component's state and manipulate that so the component's view is correctly updated, then publish my changes to the parent component. If the external prop `object` changes (let's say another component updates the object), I want those changes to overwrite what's in the internal state.

I don't doubt that there's probably an anti-pattern to the React 16.3 `getDerivedStateFromProps` working like `componentWillReceiveProps`, but I feel like that lifecycle method is still a good thing to include in future updates. I'm amiable, but I'm still just trying to understand the implications of the anti-pattern and how this new behaviour is better.

**Comment by gaearon** [comment]
@lvl99 Did you get a chance to read the [blog post](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html)?

**Comment by lvl99** [comment]
@gaearon I did, but still wanting that sweetness of the `componentWillReceiveProps` behaviour. I attempted something like (warning -- mildly untested code ahead, think of it more as pseudo-code):
```
// ...
import memoize from "memoize-one";

export default class EditObject {
  // ...
  
  getObject = memoize((propsObject, stateObject) => {
    if (propsObject && propsObject !== stateObject) {
      return propsObject
    }
    return stateObject
  })
  
  // ...
  
  render() {
    let useObject = getObject(this.props.object, this.state.object);

    // ..
  }
}
```

But the problem that I'm finding is that I just don't want to compare both objects on each render (because when  `this.state.object` changes it'll always then revert to `this.props.object`, but that's prob more a fault of my logic), just only when `this.props.object` changes. Would I perhaps use something like memoization to locally store the propsObject itself and then compare/`setState` when/if it changes? I wonder, is it "legal" to use `setState` in the memoised function too?:
```
getObject = memoize(propsObject => {
  if (propsObject !== this.state.object) {
    this.setState({
      object: propsObject
    });
    return propsObject
  }
  return this.state.object
})
```

I realise I've written a comment where I'm working stuff out at the same time. I'll modify my Codesandboxes to play around and get back to you...

**Comment by gaearon** [comment]
@lvl99 

I don't think your use case is memoization?

What you're trying to do seems to be covered here:

https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#anti-pattern-erasing-state-when-props-change

Including why it's fragile and probably not a good idea.

**Comment by lvl99** [comment]
Yeah, it's definitely not a good idea. However, just trying to figure out what the alternative is when wanting to sandbox some information that then may change due to other external side-effects. Perhaps that's a bad idea too! Just need to figure out the right path.

**Comment by gaearon** [comment]
Here's an updated example with 16.4 that works: https://codesandbox.io/s/m4km18q38j

My changes (compared to the broken version are):

* Removed unnecessary state and `getDerivedStateFromProps` from `ButtonSelectable`. It's now [fully controlled](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-controlled-component).
* Removed unnecessary `getDerivedStateFromProps` from `EditObject`. Instead of mirroring `object` in state, I separated `this.props.object` (coming from above), and `this.state.overrides` (locally edited fields). I don't attempt to "sync" them. Instead, I merge them in the render method: `{...this.props.object, this.state.overrides}`.
* I need to reset `EditObject`'s `overrides` when we change the object. However, as mentioned in the blog post, it is brittle to do so by comparing next and previous props. Even with `componentWillReceiveProps` this solution doesn't work very well (what if you need to update the object above for unrelated reasons but not reset the form state?) Instead, I added an `id` field to the objects. Now I can reset the `overrides` state using the ["uncontrolled component with a key"](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key) approach: `<EditObject key={this.state.object.id} />`.

Hope this is helpful!

**Comment by lvl99** [comment]
@gaearon very helpful! Thanks for the update and the clear breakdown. Much appreciated.

The use of the `overrides` and the `id` is very cool. Thanks for taking the time 🙏

**Comment by drcmda** [comment]
@gaearon is there anything the react team can do to warn users when they install 16.3.x? The newer gDSFP behaviour changes component fundamentals, for instance in [the transition component in react-spring](https://github.com/drcmda/react-spring/blob/master/src/Transition.js#L73-L129). Hence i get issues posted for it, for instance this one: https://github.com/drcmda/react-spring/issues/198

gDSFP is assuming it's going to be called by setState (which is used by transition when exit-out transitions are completed). As a direct result old transitions hang in there in 16.3.x until a new render pass begins.

**Comment by gaearon** [comment]
I don't think this is a serious enough bug to deprecate 16.3.x (which is the only way we can warn them). We try to use npm deprecations very sparingly — either for security vulnerabilities or for guaranteed crashes. Otherwise people won't treat them seriously.

I'm sorry that a React bug is affecting your users. One thing you could to is to cut a major release and raise the minimal required React version to 16.4. Then at least it'll be clear to new users.

**Comment by drcmda** [comment]
@gaearon could i ask for one last thing? Your answer at the top makes it seem like when libs break that's on us, not react, since our code must've been buggy already. Could you please amend this? I'm going to warn users now to upgrade react and i'm going to link to this issue. The new behaviour (which im totally fine with) is definitively changing how react functions, and if any component assumes setState is going through gDSFP it's going to break in 16.3.x, that isn't our fault, and safeguarding against that will cost some major churn. A warning would be fine, but when users come back and say "Dan Abramov said its your fault!" 😇

**Comment by gaearon** [comment]
Updated to:

>TLDR: if this change broke your code when you moved from 16.3 to 16.4, your code was already buggy in subtle ways. The change in React 16.4 made these bugs in your product code occur more often so you can fix them. We don’t consider making existing bugs in your code reproduce more reliably to be a breaking change.

>If you're coming from an issue in some third-party library that works with 16.4, but doesn't work in 16.3 — you're hitting the React bug. This issue is a complaint about this bug getting fixed in 16.4. However, we think fixing it was the right choice. We recommend everyone to upgrade to React 16.4 or later.

I think this should be clearer?

**Comment by drcmda** [comment]
nice, thanks! 🙂

**Comment by gaearon** [comment]
I'm going to lock this thread since it's super large and everyone gets notified on new comments.
If somebody has further concerns please file a new issue and we'd be happy to discuss them.
